### DISCUSSION TITLE
Using the Waveshare Pico SIM7080G with MicroPython

### DISCUSSION THREAD

**Discussion by MrTinkerman** [micropython/micropython#14356] [General]
Hi,

So I've been busy getting into LTE-M the past few days and dove in head first ordering a [Waveshare Pico SIM7080G](https://www.waveshare.com/pico-sim7080g-cat-m-nb-iot.htm)
The info I got came from the links below

First thing first, is this correct?
- The board basically has these connections: V+, GND, TX, RX and PWR_KEY to reset it.
- Communication to the module is done by sending AT commands over UART
- The board is designed for a Pico but any uc with UART and one IO should do right? (currently using a XIAO nrf52840)

I'm stuck at communicating over UART, the very basic first requirement, yay...
Maybe I broke the board by doing something stupid I'm not aware of, maybe it's faulty, maybe I'm doing something wrong.
At first I thought it was the UART [(click for topic)](https://github.com/orgs/micropython/discussions/14337)

But looking at the TX/RX on the XIAO's side, the XIAO is sending data but the RX stays low while it should be high (see screenshot below).

My basic test code is posted below too.

Does anyone have any idea what's wrong here?

Thanks!

Additional info:
- I doublecheck my soldering connections
- I power both boards from a 10C 400mAh lipo
- I tested with and without m2m simcard
- I didn't power it without the antenna connected

The links:
[How to build a cellular IoT device with the Raspberry Pi Pico — part one, the hardware](https://blog.smittytone.net/2021/08/20/how-to-build-a-cellular-iot-device-with-raspberry-pi-pico-part-one/)
[How to build a cellular IoT device with the Raspberry Pi Pico — part two, the code](https://blog.smittytone.net/2021/08/30/how-to-build-a-cellular-iot-device-with-raspberry-pi-pico-part-two/)
[korewireless/Super-Sim-Raspberry-Pi-Pico-Demos on github](https://github.com/korewireless/Super-Sim-Raspberry-Pi-Pico-Demos)
[Get Started with Super SIM SMS Commands and the Raspberry Pi Pico](https://www.twilio.com/docs/iot/supersim/get-started-with-super-sim-and-the-raspberry-pi-pico)
[Get Started with Data Comms and the Raspberry Pi Pico](https://www.twilio.com/docs/iot/supersim/get-started-with-super-sim-raspberry-pi-pico-data-comms)

![image](https://github.com/micropython/micropython/assets/16565076/71c89a03-2a1b-426e-bd43-2d957c1379be)

Test code to communicate over UART:
```py
import machine
from time import ticks_ms, sleep_ms

mdm = machine.UART(0, 115200)
pwr = machine.Pin(5, machine.Pin.OUT, machine.Pin.PULL_UP)
pwr.value(1)

'''
Info from SIM7080G_Hardware_Design_V1.05_230731.pdf

Page 26
The following methods can be used to power off SIM7080G.
Method 1: Power off SIM7080G by pulling the PWRKEY pin to ground.

Page 24
After the PWRKEY is pulled low, the module will be reset after 12.6 seconds
'''

def reset():															
                                                                        
    print('Resetting')													
                                                                        
    pwr.value(0)													
    sleep_ms(2000)   
    pwr.value(1)
        
    wait_time = 30				# Wait n seconds for the modem to become available
        
    s = ticks_ms()
    while ticks_ms() - s < wait_time*1000:
        print(f'Wait {wait_time - int((ticks_ms()-s)/1000)}')
        sleep_ms(1000)


reset()

cntr = 0

while cntr<250:				# From sstackoverflow.com/questions/64470980
                                                # The module also has auto baud rate detection, so you need to send AT a couple time to get the `OK' acknowledgement
    cntr += 1
    
    print(f'Sending AT cmd {cntr}')
    mdm.write(b'AT\r\n')
    sleep_ms(500)

    print(f'Serial data available: {mdm.any()}')
    
    if mdm.any():
        print(mdm.readline())
```

**Comment by peterhinch** [comment]
Are you sure you've connected TX on the host to RX on the modem and _vice versa_? A UART data line should never be stuck low: remaining in that state means something is broken. If you have the host RX connected to the modem RX, that would explain the stuck line.

Failing that check with a meter that the line isn't somehow strapped to gnd.

**Reply by MrTinkerman** [reply]
Yes, I have TX and RX crossed, TX to RX and RX to TX.
And no, there is no connection between the 7080's TX and ground.

So the outcome would be "something" is broken?
Either I messed up in the process or it was a DOA. My guess would be the first.

Before I order and break another one, my assumptions under "First thing first, is this correct?..."
Are those correct?

Thanks.

**Reply by peterhinch** [reply]
I would disconnect the line which is stuck low and determine which module is dragging it down.

**Reply by robert-hh** [reply]
@MrTinkerman How did you connect the PWERKEY pin? As far as I understand the documentation, it has to be pull low shortly to power up the modem and then kept high. If kept low permanently, the modem stays in reset state.

**Reply by MrTinkerman** [reply]
Ok, so first I completely disconnected the SIM7080G board and focused on my XIAO.
- I use pin 5 to control the PWR_KEY, it behaves as expected, setting it to high at startup, pulling it low for 2 seconds to reset the SIM7080G
- On startup the TX on the XIAO goes high and the RX is low

With the board connected, V+, (shared) GND and PWR_KEY I read the TX and RX from the 7080:
- Both stay low

Is the following correct:

In general, on the UART of a board:
- TX should be pulled high by the board itself and pulled low when transmitting
- The RX should remain low, only to be pulled up but the TX of the board on the other side

If the above is correct, that would mean that the 7080 is not pulling up it's TX.
And that would be wrong, right?

Thanks

**Reply by robert-hh** [reply]
Looking at the Waveshare Demo code, they treat PWR_KEY different. They set it High for 2 seconds, and then low. Maybe I recalled that wrong from the documentation. But the UART is used straight. No inversion, at 115200 baud.

**Reply by MrTinkerman** [reply]
Can I ask what Waveshare demo code you are referring to?
Because:
```
Info from SIM7080G_Hardware_Design_V1.05_230731.pdf

Page 26
The following methods can be used to power off SIM7080G.
Method 1: Power off SIM7080G by pulling the PWRKEY pin to ground.
```

And what exactly do you mean by "the UART is used straight", TX to TX and RX to RX?
I'm using the default 115200 and I'm not sure what inversion means.

Thanks.

**Reply by robert-hh** [reply]
About PWRKEY & PWR_EN: The Waveshare Breakout has an inverter at the PWR_EN pin routed to the RPI Pico connector. So PWR_EN is not directly connected to the SIM7080 PWREY. When PWR_EN is left open, PWRKEY is high.

Demo code from this web site: https://www.waveshare.com/wiki/Pico-SIM7080G-Cat-M/NB-IoT#Document

With straight I mean:  not inverted. But the wiring according to their schematics is TX <-> RX, RX <-> TX.

**Reply by MrTinkerman** [reply]
Thank you for clarifying, I will look into it tomorrow or over the weekend.

**Reply by peterhinch** [reply]
> In general, on the UART of a board:
TX should be pulled high by the board itself and pulled low when transmitting
The RX should remain low, only to be pulled up but the TX of the board on the other side

As you state, the TX idles high and pulses low when transmitting.
The RX is high impedance. If you measure with a voltmeter between RX and Gnd it will read low. If you measure with a high impedance instrument like an oscilloscope, it may show an indeterminate level. The fact that it is high impedance means that, when connected to a TX, the wire behaves as per the TX.

A TX, disconnected from other equipment, that is stuck low has a hardware problem.

**Reply by MrTinkerman** [reply]
Thanks so much for explaining.

"_A TX, disconnected from other equipment, that is stuck low has a hardware problem._"

The TX on the Pico SIM7080 stays low no matter what, also if stand alone.

Conclusion, this board is a write off :(

**Reply by radu022003** [reply]
Hello, I have tried with the same board and the AT_TEST.py example and I could observe that the modem (SIM7080) was not sending anything back. When I connected a FT232 to the RX of the pico, I could observe that the communication was working as expected. My assumption is that the level shifter (TXB0104PWR) is not OK or it needs something changed (hardware).
Does someone has an idea what should be changed? I will also write an email to waveshare.

**Reply by MrTinkerman** [reply]
I'm not yet sure of what you have tried but I am interested in the outcome/reply from Waveshare.
I'd prefer to use the Waveshare board instead of the m5 stamp I'm currently using

**Comment by Pawelinode** [comment]
You have not defined RXD and TXD pins in machine.UART(0, 115200)
Are the default values correct?

**Reply by MrTinkerman** [reply]
Yes they are correct. Long story short, on the nrf52840, adding more parameters then uart and baubrate is not supported.
I am using a https://docs.m5stack.com/en/stamp/stamp_catm now and got it working without any issues.

**Comment by CrabbyPete** [comment]
I am curious, did you ever get this to work?

**Reply by MrTinkerman** [reply]
No, I switched to the m5stack module and the waveshare module has been collecting dust since then.

**Comment by CrabbyPete** [comment]
Are you using the same code?
